### ISSUE TITLE
[🐛 Bug]: OpenQA.Selenium.WebDriverException : A WebSocket address for DevTools protocol has been detected, but the protocol version cannot be automatically detected. You must specify a protocol version.

### ISSUE THREAD

**Issue by MJB222398** [SeleniumHQ/selenium#13502]
### What happened?

I am trying to access the browser console logs across Chrome, Edge, and Firefox using the BiDi implementation. I am using a docker based Selenium Grid running on the WSL2 Ubuntu installation on my Windows PC. The C# code itself is running on this Windows machine and so I am using the RemoteWebDriver. I get this exception below when I use the method `IJavaScriptEngine.StartEventMonitoring()`:

```
Message: 
OpenQA.Selenium.WebDriverException : Unexpected error creating WebSocket DevTools session.
  ----> OpenQA.Selenium.WebDriverException : A WebSocket address for DevTools protocol has been detected, but the protocol version cannot be automatically detected. You must specify a protocol version.

  Stack Trace: 
RemoteWebDriver.GetDevToolsSession(DevToolsOptions options)
RemoteWebDriver.GetDevToolsSession()
<.ctor>b__0()
Lazy`1.ViaFactory(LazyThreadSafetyMode mode)
Lazy`1.ExecutionAndPublication(LazyHelper executionAndPublication, Boolean useDefaultConstructor)
Lazy`1.CreateValue()
JavaScriptEngine.StartEventMonitoring()
```

### How can we reproduce the issue?

```shell
internal sealed class MyDriver : IMyDriver
    {
        private readonly IList<string> _consoleLogs;
        private readonly RemoteWebDriver _webDriver;
        private readonly IJavaScriptEngine _jsEngine;

        public MyDriver(RemoteWebDriver webDriver)
        {
            _webDriver = webDriver ?? throw new ArgumentNullException(nameof(webDriver));
            _jsEngine = new JavaScriptEngine(webDriver);
            _consoleLogs = new List<string>();
        }

        public async Task Initialize()
        {
            _jsEngine.JavaScriptConsoleApiCalled += (sender, e) => _consoleLogs.Add($"{e.MessageTimeStamp}: {e.MessageType} - {e.MessageContent}");
            await _jsEngine.StartEventMonitoring();
        }
    }
```
```


### Relevant log output

```shell
13:01:03.963 INFO [LocalNode.newSession] - Session created by the Node. Id: 286084821c57eed09a05fef9ecd69ac4, Caps: Capabilities {acceptInsecureCerts: false, browserName: chrome, browserVersion: 120.0.6099.224, chrome: {chromedriverVersion: 120.0.6099.109 (3419140ab66..., userDataDir: /home/seluser/selenium/profile}, fedcm:accounts: true, goog:chromeOptions: {debuggerAddress: localhost:35435}, goog:loggingPrefs: {browser: ALL, client: ALL, driver: ALL, performance: ALL, profiler: ALL, server: ALL}, networkConnectionEnabled: false, pageLoadStrategy: normal, platformName: linux, proxy: Proxy(), se:bidiEnabled: false, se:cdp: ws://172.20.0.4:4444/sessio..., se:cdpVersion: 120.0.6099.224, se:vnc: ws://172.20.0.4:4444/sessio..., se:vncEnabled: true, se:vncLocalAddress: ws://172.20.0.4:7900, setWindowRect: true, strictFileInteractability: false, timeouts: {implicit: 0, pageLoad: 300000, script: 30000}, unhandledPromptBehavior: dismiss and notify, webauthn:extension:credBlob: true, webauthn:extension:largeBlob: true, webauthn:extension:minPinLength: true, webauthn:extension:prf: true, webauthn:virtualAuthenticators: true}
```

### Operating System

Windows 10

### Selenium version

.NET Selenium 4.17

### What are the browser(s) and version(s) where you see this issue?

Chrome: 120.0.6099.224, Edge: 120.0.2210.144 

### What are the browser driver(s) and version(s) where you see this issue?

Chrome: 120.0.6099.224, Edge: 120.0.2210.144 

### Are you using Selenium Grid?

Yes 4.17.0-20240123

**Comment by diemol** [comment]
How do you start the Docker container?

**Comment by MJB222398** [comment]
> How do you start the Docker container?

I use docker-compose. So `docker-compose up -d` ran from the Ubuntu terminal

**Comment by MJB222398** [comment]
> How do you start the Docker container?

@diemol  I use docker-compose. So `docker-compose up -d` ran from the Ubuntu terminal

**Comment by diemol** [comment]
OK. But how does the compose file look like?

**Comment by MJB222398** [comment]
> OK. But how does the compose file look like?

@diemol 

```
# To execute this docker-compose yml file use `docker-compose -f docker-compose.yml up`
# Add the `-d` flag at the end for detached execution
# To stop the execution, hit Ctrl+C, and then `docker-compose -f docker-compose.yml down`
version: "3"
services:
  chrome:
    image: selenium/node-chrome:4.17.0-20240123
    container_name: chrome-node
    shm_size: 2gb
    privileged: true
    depends_on:
      - selenium-hub
    environment:
      - HUB_HOST=selenium-hub
      - SE_EVENT_BUS_HOST=selenium-hub
      - SE_EVENT_BUS_PUBLISH_PORT=4442
      - SE_EVENT_BUS_SUBSCRIBE_PORT=4443
      - SE_NODE_MAX_SESSIONS=1
    ports:
      - "6900:5900"
    volumes:
      - ./zscaler:/zscaler
      - ./chrome:/selenium/chrome
    command:
      - /bin/bash
      - -c
      - |
        sh /zscaler/zscaler.sh
        sh /selenium/chrome/profile-setup.sh
        sh /opt/bin/entry_point.sh
    networks:
      - selenium_grid_internal

  edge:
    image: selenium/node-edge:4.17.0-20240123
    container_name: edge-node
    shm_size: 2gb
    privileged: true
    depends_on:
      - selenium-hub
    environment:
      - HUB_HOST=selenium-hub
      - SE_EVENT_BUS_HOST=selenium-hub
      - SE_EVENT_BUS_PUBLISH_PORT=4442
      - SE_EVENT_BUS_SUBSCRIBE_PORT=4443
      - SE_NODE_MAX_SESSIONS=1
    ports:
      - "6901:5900"
    volumes:
      - ./zscaler:/zscaler
      - ./edge:/selenium/edge
    command:
      - /bin/bash
      - -c
      - |
        sh /zscaler/zscaler.sh
        sh /selenium/edge/profile-setup.sh
        sh /opt/bin/entry_point.sh
    networks:
      - selenium_grid_internal

  firefox:
    image: selenium/node-firefox:4.17.0-20240123
    container_name: firefox-node
    shm_size: 2gb
    privileged: true
    depends_on:
      - selenium-hub
    environment:
      - HUB_HOST=selenium-hub
      - SE_EVENT_BUS_HOST=selenium-hub
      - SE_EVENT_BUS_PUBLISH_PORT=4442
      - SE_EVENT_BUS_SUBSCRIBE_PORT=4443
      - SE_NODE_MAX_SESSIONS=1
    ports:
      - "6902:5900"
    volumes:
      - ./zscaler:/zscaler
      - ./firefox:/selenium/firefox
    command:
      - /bin/bash
      - -c
      - |
        sh /zscaler/zscaler.sh
        sh /selenium/firefox/profile-setup.sh
        sh /opt/bin/entry_point.sh
    networks:
      - selenium_grid_internal

  chrome_video:
    image: selenium/video:ffmpeg-6.1-20240123
    privileged: true
    volumes:
      - ${VIDEOS_DIRECTORY:-./videos}:/videos
    depends_on:
      - chrome
    environment:
      - DISPLAY_CONTAINER_NAME=chrome
      - FILE_NAME=chrome_video.mp4
    networks:
      - selenium_grid_internal

  edge_video:
    image: selenium/video:ffmpeg-6.1-20240123
    privileged: true
    volumes:
      - ${VIDEOS_DIRECTORY:-./videos}:/videos
    depends_on:
      - edge
    environment:
      - DISPLAY_CONTAINER_NAME=edge
      - FILE_NAME=edge_video.mp4
    networks:
      - selenium_grid_internal

  firefox_video:
    image: selenium/video:ffmpeg-6.1-20240123
    privileged: true
    volumes:
      - ${VIDEOS_DIRECTORY:-./videos}:/videos
    depends_on:
      - firefox
    environment:
      - DISPLAY_CONTAINER_NAME=firefox
      - FILE_NAME=firefox_video.mp4
    networks:
      - selenium_grid_internal

  selenium-hub:
    image: selenium/hub:4.17.0-20240123
    privileged: true
    container_name: selenium-hub
    ports:
      - "4442:4442"
      - "4443:4443"
      - "4444:4444"
    environment:
      - GRID_TIMEOUT=120000
      - GRID_BROWSER_TIMEOUT=120000
    networks:
      - selenium_grid_internal

networks:
  selenium_grid_internal:
```

**Comment by diemol** [comment]
You need to set the Grid URL to do CDP/BiDi with those containers https://github.com/SeleniumHQ/docker-selenium#grid-url-and-session-timeout

**Comment by MJB222398** [comment]
> You need to set the Grid URL to do CDP/BiDi with those containers https://github.com/SeleniumHQ/docker-selenium#grid-url-and-session-timeout

@diemol Thanks, so I have added a hostname for my hub:

`hostname: selenium-hub`

And then set the SE_NODE_GRID_URL for each of the three nodes:

` - SE_NODE_GRID_URL=http://selenium-hub:4444`

but the error I am getting is the same.

I tried also with ` - SE_NODE_GRID_URL=http://localhost:4444` as well but same exception still

**Comment by nvborisenko** [comment]
It is regression issue, thanks for reporting it.

Details: `RemoteWebDriver` doesn't pass `se:cdpVersion` capability to the underlying `DevToolsSession` ctor. The fix is easy.

**Comment by MJB222398** [comment]
> It is regression issue, thanks for reporting it.
> 
> Details: `RemoteWebDriver` doesn't pass `se:cdpVersion` capability to the underlying `DevToolsSession` ctor. The fix is easy.

Thanks @nvborisenko , so the above reported error is what I get on Chrome and Edge. But actually on Firefox I just checked and there I get this instead:

`OpenQA.Selenium.WebDriverException : Cannot find se:cdpVersion capability for driver`

I guess this is just a different symptom of the same bug?

**Comment by nvborisenko** [comment]
@MJB222398 seems different issue. Can you please check with Selenium 4.16.2 to understand whether it is regression?

**Comment by MJB222398** [comment]
> @MJB222398 seems different issue. Can you please check with Selenium 4.16.2 to understand whether it is regression?

I am getting that same exception on Firefox with 4.16.2 as well (using the same container versions as in the docker-compose above)

**Comment by MJB222398** [comment]
> @MJB222398 seems different issue. Can you please check with Selenium 4.16.2 to understand whether it is regression?

@nvborisenko  I am getting that same exception on Firefox with 4.16.2 as well (using the same container versions as in the docker-compose above)

**Comment by titusfortner** [comment]
The CDP version should be hard coded to 85 for Firefox

**Comment by nvborisenko** [comment]
Right, I will also fix it. Thanks.

**Comment by nvborisenko** [comment]
@titusfortner I would like to make a patch for this (ideally all regression issues should be patched). I verified the fix for chrome/edge/firefox.

PS: for firefox we have to specify `EnableDevToolsProtocol` in firefox options.

**Comment by MJB222398** [comment]
@nvborisenko thanks for your work on getting this resolved. Is there any update on when the patch might be released?

**Comment by nvborisenko** [comment]
It is even not merged into `trunk`. I have no ideas what is a process to release a patch. @titusfortner might be you know?

**Comment by titusfortner** [comment]
once it is merged into trunk, if everything else is good, we can walk through the patch release process. I think the only newly tricky piece will be making sure we have the right Selenium Manager.

**Comment by titusfortner** [comment]
Well, only tricky because what we have now isn't versioned correctly, so we could fix it if we wanted.

**Comment by MJB222398** [comment]
> Right, I will also fix it. Thanks.

@nvborisenko I have upgraded to the new 4.18.1 selenium package and am still getting that error on Firefox. Should I raise a new ticket or should it be resolved under this?

**Comment by nvborisenko** [comment]
As I mentioned especially for Firefox it should be enabled explicitly like:

```csharp
var firefoxOptions = new FirefoxOptions
{
    EnableDevToolsProtocol = true,
};

var driver = new RemoteWebDriver(firefoxOptions); 
```

**Comment by MJB222398** [comment]
@nvborisenko  Ahh I did not see that, only the comment where you said you would fix it, my apologies.

I have made that change and I do now no longer get that error with Firefox. However it does not actually allow me to retrieve the browser console logs. For Chrome and Edge this works fine and I can retrieve but for Firefox no, there are no logs picked up. Perhaps I need to raise a new ticket for that though
